selfservice/strategy/code/code_recovery.go

```diff
@@ -0,0 +1,125 @@
+package code
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/gofrs/uuid"
+
+	"github.com/ory/herodot"
+	"github.com/ory/x/randx"
+
+	"github.com/ory/kratos/identity"
+)
+
+type RecoveryCodeType int
+
+const (
+	RecoveryCodeTypeAdmin RecoveryCodeType = iota + 1
+	RecoveryCodeTypeSelfService
+)
+
+var (
+	ErrCodeNotFound    = herodot.ErrNotFound.WithReasonf("unknown recovery code")
+	ErrCodeAlreadyUsed = herodot.ErrNotFound.WithReasonf("recovery code was already used")
+)
+
+type RecoveryCode struct {
+	// ID represents the code's unique ID.
+	//
+	// required: true
+	// type: string
+	// format: uuid
+	ID uuid.UUID `json:"id" db:"id" faker:"-"`
+
+	// Code represents the recovery code
+	Code string `json:"-" db:"code"`
+
+	// UsedAt is the timestamp of when the code was used or null if it wasn't yet
+	UsedAt sql.NullTime `json:"-" db:"used_at"`
+
+	// RecoveryAddress links this code to a recovery address.
+	// required: true
+	RecoveryAddress *identity.RecoveryAddress `json:"recovery_address" belongs_to:"identity_recovery_addresses" fk_id:"RecoveryAddressID"`
+
+	// CodeType is the type of the code - either "admin" or "selfservice"
+	CodeType RecoveryCodeType `json:"-" faker:"-" db:"code_type"`
+
+	// ExpiresAt is the time (UTC) when the code expires.
+	// required: true
+	ExpiresAt time.Time `json:"expires_at" faker:"time_type" db:"expires_at"`
+
+	// IssuedAt is the time (UTC) when the code was issued.
+	// required: true
+	IssuedAt time.Time `json:"issued_at" faker:"time_type" db:"issued_at"`
+
+	// CreatedAt is a helper struct field for gobuffalo.pop.
+	CreatedAt time.Time `json:"-" faker:"-" db:"created_at"`
+	// UpdatedAt is a helper struct field for gobuffalo.pop.
+	UpdatedAt time.Time `json:"-" faker:"-" db:"updated_at"`
+	// RecoveryAddressID is a helper struct field for gobuffalo.pop.
+	RecoveryAddressID uuid.NullUUID `json:"-" faker:"-" db:"identity_recovery_address_id"`
+	// FlowID is a helper struct field for gobuffalo.pop.
+	FlowID     uuid.UUID `json:"-" faker:"-" db:"selfservice_recovery_flow_id"`
+	NID        uuid.UUID `json:"-" faker:"-" db:"nid"`
+	IdentityID uuid.UUID `json:"identity_id" faker:"-" db:"identity_id"`
+}
+
+func (RecoveryCode) TableName(ctx context.Context) string {
+	return "identity_recovery_codes"
+}
+
+func (f RecoveryCode) IsExpired() bool {
+	return f.ExpiresAt.Before(time.Now())
+}
+
+func (r RecoveryCode) WasUsed() bool {
+	return r.UsedAt.Valid
+}
+
+func GenerateRecoveryCode() string {
+	return randx.MustString(8, randx.Numeric)
+}
+
+type RecoveryCodeDTO struct {
```

Comment:
So this struct is only used internally by the persister right? Why does it not have `db` tags? Why does the other struct have them?

Reply:
This struct is only used as a Data Transfer Object, meaning we are populating its values with the stuff we know when creating a new code. Then it is passed into the persister method, which then "transforms" it into the _actual_ entity by taking all the stuff from this struct into the new `RecoveryCode` object. This struct does not represent the entity, but is rather just "parameters" to a method.

>Why does the other struct have them?

Because `RecoveryCode` is what actually represents the entity.

Reply:
Looking at the code, it does not seem to be neccessary to have this object. The fields are all the same data types as far as I can tell. That means that we can do the transformations on the `RecoveryCode` struct itself and don't need this indirection.

If you were to have other or less fields, the struct would be called something `CreateRecoveryCodeParams` :)

Reply:
Ok, then maybe call it `RawCode` or `CodePlaintext` to make it clear in `CreateRecoveryCodeParams`

Reply:
Alright, I named it `CodeHMAC` on the `RecoveryCode` and `RawCode` on `CreateRecoveryCodeParams`.

Reply:
👍